### PR TITLE
feat: support table joins on key format mismatch

### LINKED ISSUE #6229: [Auto repartitioning support]: Auto repartition joins on key format mismatch
**Auto-repartitioning on key format mismatch**. Adds support for automatic repartitioning of streams and tables for joins where key formats do not match.
1. Update the logical plan so that we introduce repartition steps on (right) sources as needed to ensure correct key format.
    - Make use of existing repartitions where possible.
1. Update docs to cover 
    - the implicit / automatic repartitioning. 
     - repartition the _right_ source and _why_ (see KLIP-33 for why).  For stream-stream and table-table joins, use can control which side is repartitioned by changing sides.
     - example of how to manually repartition a source to avoid implicit repartitioning, e.g. stream-table join where the order can be flipped.
1. Lots of QTT tests!
1. Doc updates.

### PR DESCRIPTION
**Pull request by agavra**
### Description 

stacked on #6609 - which will not be merged until this PR is approved, don't review first two commits!

fixes #6229 and also remove the code around "required formats". This means that we will always chose the first "preferred" format using the algoirthm outlined in #6393 but removing the restriction that tables are not re-partitioned. 

### Testing done 

QTT testing

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json [hunk 1]
```diff
@@ -1227,41 +1238,41 @@
       ],
       "post": {
         "topics": {
-          "topics" : [
+          "topics": [
             {
-              "name" : "_confluent-ksql-some.ksql.service.idquery_CSAS_OUTPUT_0-L_Join-left-repartition",
-              "keyFormat" : {"format" : "JSON_SR", "features": ["UNWRAP_SINGLES"]},
-              "valueFormat" : {"format" : "JSON"}
+              "name": "_confluent-ksql-some.ksql.service.idquery_CSAS_OUTPUT_0-KafkaTopic_Right-Reduce-changelog",
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json:1243]: This topic should also go away with https://github.com/confluentinc/ksql/issues/6650 -- it makes sense that all internal topics should be Avro formatted in this case.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json [hunk 2]
```diff
@@ -1227,41 +1238,41 @@
       ],
       "post": {
         "topics": {
-          "topics" : [
+          "topics": [
```
Comment (vcrfxia) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/multi-joins.json:1241]: nit: can we order these topics in a more readable fashion? Previously it was repartition topics first, then changelogs, ordered from left to right.

**Review by vcrfxia — APPROVED** [review]
Awesome -- thanks @agavra ! A great simplification to the join key format resolution logic.

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json [hunk 3]
```diff
@@ -2250,14 +2250,20 @@
         "topics" : {
           "topics" : [
             {
-              "name" : "_confluent-ksql-some.ksql.service.idquery_CSAS_OUTPUT_0-Join-repartition",
-              "keyFormat" : {"format" : "KAFKA"},
-              "valueFormat" : {"format" : "JSON"}
+              "name": "_confluent-ksql-some.ksql.service.idquery_CSAS_OUTPUT_0-RightSourceKeyed-SelectKey-repartition",
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-functional-tests/src/test/resources/query-validation-tests/joins.json:2253]: yes, it is backwards incompatible in that sense. it's not in the sense that all old enqueued commands would still work in the same way. I will call that out

File: ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java [hunk 4]
```diff
@@ -112,48 +112,18 @@ public JoinNode(
...
-   * <p>For now, the left key format is the preferred join key format unless either:
-   * <ul>
-   *   <li>The right source is not already being repartitioned and the left source is.</li>
-   *   <li>The right source is a table.</li>
-   * </ul>
-   * In which case, the right key format it used.
-   *
-   * <p>An exception is currently thrown if both sides are tables and their key formats differ.
+   * <p>For now, the left key format is the preferred join key format unless the
+   * right source is not already being repartitioned and the left source is.
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:116]: From KLIP-33:

> Where one side must be repartitioned to correct the key format, choosing which side to repartition 
can not be driven by the size of the data, as in a traditional database system, as the size of 
the data is unknown, likely infinite. Ideally, for a streaming system it is the rate of change of 
the data, i.e. the throughput, that would drive the choice. Unfortunately, this too can not be 
known upfront.  For this reason, we propose repartitioning based on the order of sources within 
the query, with the source on the _right_ being repartitioned.

I'm still choosing to repartition the _left_ source, not the right.

[the remark this answers is not in the document]
Comment (vcrfxia) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:116]: The changes in this PR chose the preferred format from the left source, which means the right source is being repartitioned. The QTTs support this as well. Have I missed something?

[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/JoinNode.java:116]: 😬 yes... you are right!

**Pull request merged**